### PR TITLE
feat: select, handle and ls

### PR DESCRIPTION
**Pull request by alanshaw**
This PR supersedes #1 and #2 and adds `ls` and `handle` on top of `select` and adds 100% test coverage:

Screenshot 2019-08-14 at 12 21 52

API has changed a little from #2 because we actually do need to keep some state around for `ls`.

If you use `ls` you need to remember you've done the multistream-select handshake so that you can call `select` afterwards.

So the API is now:

```js
const dialer = new MSS.Dialer(muxedStream)
const protocols = await dialer.ls() // optional!
const { stream, protocol } = await dialer.select('/echo/1.0.0')

// --- and ---

const listener = new MSS.Listener(muxedStream)
const { stream, protocol } = await listener.handle('/echo/1.0.0')
```

This work also uncovered a possible bug in `go-multistream`  https://github.com/multiformats/go-multistream/issues/41

I also addressed feedback from #2 here.

If this is approved I'll open a PR to `js-multistream-select`.

closes #1
closes #2

cc @jacobheun @vasco-santos @daviddias

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    // mss.handle(handledProtocols)
+    // Returns selected stream and protocol
+    const { stream, protocol } = await mss.handle([
+      '/ipfs-dht/1.0.0',
+      '/ipfs-bitswap/1.0.0'
+    ])
+
+    // Typically here we'd call the handler function that was registered in
+    // libp2p for the given protocol:
+    // e.g. handlers[protocol].hander(stream)
```
Comment (jacobheun) [inline README.md]:
```suggestion
    // e.g. handlers[protocol].handler(stream)
```

File: src/handle.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  protocols = Array.isArray(protocols) ? protocols : [protocols]
+  const { reader, writer, rest } = toReaderWriter(stream)
+
+  while (true) {
+    const protocol = (await Multistream.read(reader)).toString()
+    log('read "%s"', protocol)
+
+    if (protocols.includes(protocol)) {
+      Multistream.write(writer, protocol)
+      log('write "%s" "%s"', protocol, protocol)
```
Comment (jacobheun) [inline src/handle.js:20]: Did you mean to log something other than the protocol? It's being logged twice here.
Oh, perhaps this is saying "writing on 'protocol', 'protocol'", as this seems to be the format of the other logs.

File: test/integration.spec.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  it('should handle, ls and select', async () => {
+    const protocols = ['/echo/2.0.0', '/echo/1.0.0']
+    const selectedProtocol = protocols[protocols.length - 1]
+    const pair = DuplexPair()
+
+    const dialer = new MSS.Dialer(pair[0])
+    const listener = new MSS.Listener(pair[1])
+
+    const [dialerSelection, listenerSelection] = await Promise.all([
```
Comment (jacobheun) [inline test/integration.spec.js]: These look to be flipped
```suggestion
    const [listenerSelection, dialerSelection] = await Promise.all([
```

**Review by jacobheun — APPROVED** [review]
Just a couple nits, but otherwise this looks good to me, great work! 🎉 

Just a note on `ls`, this is typically run in parallel (both js and go) on another stream as to not block operations, so you won't often see the select happen afterwards on the same stream. If people are running non muxed streams (resource restricted environments), the test scenarios you've added will be really helpful (ls before select).

For go, I think the delimited writer takes care of some of this, but I'm not positive. In most cases I believe we just negotiate blindly anyway, so it's entirely possible there could be a bug there.

File: src/handle.js (hunk 2 above)
Comment (alanshaw) [reply src/handle.js:20]: Yes exactly!

**Comment by alanshaw** [comment]
> Just a note on `ls`, this is typically run in parallel (both js and go) on another stream as to not block operations, so you won't often see the select happen afterwards on the same stream. If people are running non muxed streams (resource restricted environments), the test scenarios you've added will be really helpful (ls before select).

Ah, super interesting, I did not consider that...well, this supports both ways now so 🤷‍♂️

**Pull request merged**